File: UnionFind/UnionFind.py
class UnionFind:
    def __init__(self, n):
        self.parent = [-1] * n
        self.rank = [1] * n
    
    def find(self, x):
        if self.parent[x] == -1: return x
        self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def unite(self, x, y):
        x = self.find(x)
        y = self.find(y)
        if x == y: return False
        if self.rank[x] > self.rank[y]: x, y = y, x
        if self.rank[x] == self.rank[y]: self.rank[y] += 1
        self.parent[x] = y
        return True
    
    def issame(self, x, y): 
        return self.find(x) == self.find(y)

File: UnionFind/UnionFind.cpp
#include<bits/stdc++.h>
using namespace std;

struct UnionFind {
    private:
    int *parent, *rank;
    public:
    UnionFind(int n) {
        parent = (int*)malloc(n*sizeof(int));
        rank = (int*)malloc(n*sizeof(int));
        fill_n(parent, n, -1);
        fill_n(rank, n, 1);
    }
    int find(int x) {
        if (parent[x] == -1) return x;
        parent[x] = find(parent[x]);
        return parent[x];
    }
    bool unite(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) return false;
        if (rank[x] > rank[y]) swap(x, y);
        if (rank[x] == rank[y]) rank[y] += 1;
        parent[x] = y;
        return true;
    }
    bool issame(int x, int y) {
        return find(x) == find(y);
    }
    ~UnionFind() {
        free(parent);
        free(rank);
    }
};